In [1]:
# Import required libraries and dependencies
import pandas as pd
import hvplot.pandas
from sklearn.cluster import KMeans
from sklearn.decomposition import PCA
from sklearn.preprocessing import StandardScaler

In [2]:
# Load the data into a Pandas DataFrame
df_market_data = pd.read_csv(
    "Resources/crypto_market_data.csv",
    index_col="coin_id")

# Display sample data
df_market_data.head(10)

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                    
bitcoin                              1.08388                     7.60278   
ethereum                             0.22392                    10.38134   
tether                              -0.21173                     0.04935   
ripple                              -0.37819                    -0.60926   
bitcoin-cash                         2.90585                    17.09717   
binancecoin                          2.10423                    12.85511   
chainlink                           -0.23935                    20.69459   
cardano                              0.00322                    13.99302   
litecoin                            -0.06341                     6.60221   
bitcoin-cash-sv                      0.92530                     3.29641   

                 price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                     
bitcoin                              6.57509                      7.67258   
ethereum                             4.80849                      0.13169   
tether                               0.00640                     -0.04237   
ripple                               2.24984                      0.23455   
bitcoin-cash                        14.75334                     15.74903   
binancecoin                          6.80688                      0.05865   
chainlink                            9.30098                    -11.21747   
cardano                              5.55476                     10.10553   
litecoin                             7.28931                      1.21662   
bitcoin-cash-sv                     -1.86656                      2.88926   

                 price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                      
bitcoin                             -3.25185                      83.51840   
ethereum                           -12.88890                     186.77418   
tether                               0.28037                      -0.00542   
ripple                             -17.55245                      39.53888   
bitcoin-cash                       -13.71793                      21.66042   
binancecoin                         36.33486                     155.61937   
chainlink                          -43.69522                     403.22917   
cardano                            -22.84776                     264.51418   
litecoin                           -17.23960                      27.49919   
bitcoin-cash-sv                    -24.87434                       7.42562   

                 price_change_percentage_1y  
coin_id                                      
bitcoin                            37.51761  
ethereum                          101.96023  
tether                              0.01954  
ripple                            -16.60193  
bitcoin-cash                       14.49384  
binancecoin                        69.69195  
chainlink                         325.13186  
cardano                           156.09756  
litecoin                          -12.66408  
bitcoin-cash-sv                    93.73082

In [3]:
# Generate summary statistics
df_market_data.describe()

price_change_percentage_24h  price_change_percentage_7d  \
count                    41.000000                   41.000000   
mean                     -0.269686                    4.497147   
std                       2.694793                    6.375218   
min                     -13.527860                   -6.094560   
25%                      -0.608970                    0.047260   
50%                      -0.063410                    3.296410   
75%                       0.612090                    7.602780   
max                       4.840330                   20.694590   

       price_change_percentage_14d  price_change_percentage_30d  \
count                    41.000000                    41.000000   
mean                      0.185787                     1.545693   
std                       8.376939                    26.344218   
min                     -18.158900                   -34.705480   
25%                      -5.026620                   -10.438470   
50%                       0.109740                    -0.042370   
75%                       5.510740                     4.578130   
max                      24.239190                   140.795700   

       price_change_percentage_60d  price_change_percentage_200d  \
count                    41.000000                     41.000000   
mean                     -0.094119                    236.537432   
std                      47.365803                    435.225304   
min                     -44.822480                     -0.392100   
25%                     -25.907990                     21.660420   
50%                      -7.544550                     83.905200   
75%                       0.657260                    216.177610   
max                     223.064370                   2227.927820   

       price_change_percentage_1y  
count                   41.000000  
mean                   347.667956  
std                   1247.842884  
min                    -17.567530  
25%                      0.406170  
50%                     69.691950  
75%                    168.372510  
max                   7852.089700

In [4]:
# Plot your data to see what's in your DataFrame
df_market_data.hvplot.line(
    width=800,
    height=400,
    rot=90
)

:NdOverlay   [Variable]
   :Curve   [coin_id]   (value)

---

### Prepare the Data

In [5]:
# Use the `StandardScaler()` module from scikit-learn to normalize the data from the CSV file

# Create the StandardScaler instance
scaler = StandardScaler()
    
# Fit the scaler and transform the data
scaled_data = scaler.fit_transform(df_market_data)

In [6]:
# Create a DataFrame with the scaled data

scaled_df = pd.DataFrame(
        scaled_data,
        columns=df_market_data.columns,
        index=df_market_data.index
    )
# Copy the crypto names from the original DataFrame


# Set the coin_id column as index


# Display the scaled DataFrame
scaled_df.head()


price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                 
bitcoin                          0.508529                    0.493193   
ethereum                         0.185446                    0.934445   
tether                           0.021774                   -0.706337   
ripple                          -0.040764                   -0.810928   
bitcoin-cash                     1.193036                    2.000959   

              price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                  
bitcoin                          0.772200                     0.235460   
ethereum                         0.558692                    -0.054341   
tether                          -0.021680                    -0.061030   
ripple                           0.249458                    -0.050388   
bitcoin-cash                     1.760610                     0.545842   

              price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                   
bitcoin                         -0.067495                     -0.355953   
ethereum                        -0.273483                     -0.115759   
tether                           0.008005                     -0.550247   
ripple                          -0.373164                     -0.458259   
bitcoin-cash                    -0.291203                     -0.499848   

              price_change_percentage_1y  
coin_id                                   
bitcoin                        -0.251637  
ethereum                       -0.199352  
tether                         -0.282061  
ripple                         -0.295546  
bitcoin-cash                   -0.270317

---

### Find the Best Value for k Using the Original Scaled DataFrame.

In [7]:
# Create a list with the number of k-values from 1 to 11
k_range = list(range(1,12))

In [8]:
# Create an empty list to store the inertia values
scaled_inertias = []

# Create a for loop to compute the inertia with each possible value of k
# Inside the loop:
# 1. Create a KMeans model using the loop counter for the n_clusters
# 2. Fit the model to the data using `df_market_data_scaled`
# 3. Append the model.inertia_ to the inertia list
for k in k_range:
        kmeans = KMeans(n_clusters=k, random_state=0)
        kmeans.fit(scaled_df)
        scaled_inertias.append(kmeans.inertia_)
    


In [9]:
# Create a dictionary with the data to plot the Elbow curve
# Create the elbow curve
elbow_data = {
    'k': list(k_range),
    'inertia': scaled_inertias,
    'Dataset': 'Scaled Data'
    }

# Create a DataFrame with the data to plot the Elbow curve
elbow_df = pd.DataFrame(elbow_data)

In [10]:
# Plot a line chart with all the inertia values computed with
# the different values of k to visually identify the optimal value for k.
elbow_df.hvplot.line(
        x='k', 
        y='inertia',
        title='Elbow Curve',
        xticks=k_range
    )


:Curve   [k]   (inertia)

#### Answer the following question: 

**Question:** What is the best value for `k`?

**Answer:** 4


---

### Cluster Cryptocurrencies with K-means Using the Original Scaled DataFrame

In [11]:
# Initialize the K-Means model using the best value for k
k=4
kmeans = KMeans(n_clusters=k, random_state=0)


In [12]:
# Fit the K-Means model using the scaled DataFrame
kmeans.fit(scaled_df)

KMeans(n_clusters=4, random_state=0)

In [13]:
# Predict the clusters to group the cryptocurrencies using the scaled DataFrame
clusters = kmeans.predict(scaled_df)
# Print the resulting array of cluster values.
print("Cluster predictions:")
print(clusters)

Cluster predictions:
[2 2 0 0 2 2 2 2 2 0 0 0 0 2 0 2 0 0 2 0 0 2 0 0 0 0 0 0 2 0 0 0 1 2 0 0 3
 0 0 0 0]


In [14]:
# Create a copy of the scaled DataFrame
clustered_df = scaled_df.copy()

In [15]:
# Add a new column to the copy of the scaled DataFrame with the predicted clusters
clustered_df['CryptoCluster'] = clusters
clustered_df['Dataset'] = 'Scaled Data'
# Display the copy of the scaled DataFrame
print("\nClustered DataFrame:")
clustered_df.head()


Clustered DataFrame:


price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                 
bitcoin                          0.508529                    0.493193   
ethereum                         0.185446                    0.934445   
tether                           0.021774                   -0.706337   
ripple                          -0.040764                   -0.810928   
bitcoin-cash                     1.193036                    2.000959   

              price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                  
bitcoin                          0.772200                     0.235460   
ethereum                         0.558692                    -0.054341   
tether                          -0.021680                    -0.061030   
ripple                           0.249458                    -0.050388   
bitcoin-cash                     1.760610                     0.545842   

              price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                   
bitcoin                         -0.067495                     -0.355953   
ethereum                        -0.273483                     -0.115759   
tether                           0.008005                     -0.550247   
ripple                          -0.373164                     -0.458259   
bitcoin-cash                    -0.291203                     -0.499848   

              price_change_percentage_1y  CryptoCluster      Dataset  
coin_id                                                               
bitcoin                        -0.251637              2  Scaled Data  
ethereum                       -0.199352              2  Scaled Data  
tether                         -0.282061              0  Scaled Data  
ripple                         -0.295546              0  Scaled Data  
bitcoin-cash                   -0.270317              2  Scaled Data

In [37]:
# Create a scatter plot using hvPlot by setting
# `x="price_change_percentage_24h"` and `y="price_change_percentage_7d"`.
# Color the graph points with the labels found using K-Means and
# add the crypto name in the `hover_cols` parameter to identify
# the cryptocurrency represented by each data point.
scaled_scatter_plot = clustered_df.hvplot.scatter(
    x="price_change_percentage_24h",
    y="price_change_percentage_7d",
    by="CryptoCluster",
    hover_cols=["coin_id"],
    title="Scaled Cryptocurrency Clusters"
)
scaled_scatter_plot

:NdOverlay   [CryptoCluster]
   :Scatter   [price_change_percentage_24h]   (price_change_percentage_7d,coin_id)

---

### Optimize Clusters with Principal Component Analysis.

In [38]:
# Create a PCA model instance and set `n_components=3`.
pca = PCA(n_components=3)

In [39]:
# Use the PCA model with `fit_transform` to reduce the original scaled DataFrame
# down to three principal components.
pca_data = pca.fit_transform(scaled_df)
# View the scaled PCA data


In [40]:
# Retrieve the explained variance to determine how much information
# can be attributed to each principal component.
explained_variance = pca.explained_variance_ratio_
# Print explained variance for each component
print("Explained Variance Ratio by Principal Component:")
for i, variance in enumerate(explained_variance):
    print(f"PC {i+1}: {variance:.4f} ({variance*100:.2f}%)")

# Calculate and print total explained variance
total_variance = explained_variance.sum()
print(f"\nTotal Explained Variance: {total_variance:.4f} ({total_variance*100:.2f}%)")

Explained Variance Ratio by Principal Component:
PC 1: 0.3720 (37.20%)
PC 2: 0.3470 (34.70%)
PC 3: 0.1760 (17.60%)

Total Explained Variance: 0.8950 (89.50%)


#### Answer the following question: 

**Question:** What is the total explained variance of the three principal components?

**Answer:**  89.50%



In [41]:
# Create a new DataFrame with the PCA data.
pca_columns = ['PC1','PC2', 'PC3']
pca_df = pd.DataFrame(
    data=pca_data,
    columns=pca_columns,
    index=scaled_df.index
)

# Copy the crypto names from the original scaled DataFrame


# Set the coin_id column as index


# Display the scaled PCA DataFrame
pca_df.head()

PC1       PC2       PC3
coin_id                                   
bitcoin      -0.600667  0.842760  0.461595
ethereum     -0.458261  0.458466  0.952877
tether       -0.433070 -0.168126 -0.641752
ripple       -0.471835 -0.222660 -0.479053
bitcoin-cash -1.157800  2.041209  1.859715

---

### Find the Best Value for k Using the Scaled PCA DataFrame

In [42]:
# Create a list with the number of k-values from 1 to 11
k_range = list(range(1,12))

In [43]:
# Create an empty list to store the inertia values
pca_inertias = []

# Create a for loop to compute the inertia with each possible value of k
# Inside the loop:
# 1. Create a KMeans model using the loop counter for the n_clusters
# 2. Fit the model to the data using `df_market_data_pca`
# 3. Append the model.inertia_ to the inertia list
for k in k_range:
        kmeans = KMeans(n_clusters=k, random_state=0)
        kmeans.fit(pca_df)
        pca_inertias.append(kmeans.inertia_)

In [44]:
# Create a dictionary with the data to plot the Elbow curve
elbow_data = {
    'k': list(k_range),
    'inertia': pca_inertias,
    'Dataset': 'PCA Data'
    }

# Create a DataFrame with the data to plot the Elbow curve
elbow_df_pca = pd.DataFrame(elbow_data)

In [45]:
# Plot a line chart with all the inertia values computed with
# the different values of k to visually identify the optimal value for k.
elbow_df_pca.hvplot.line(
        x='k', 
        y='inertia',
        title='Elbow Curve',
        xticks=k_range
    )


:Curve   [k]   (inertia)

#### Answer the following questions: 

* **Question:** What is the best value for `k` when using the PCA data?

  * **Answer:** 4


* **Question:** Does it differ from the best k value found using the original data?

  * **Answer:** Slightly, the inertia values are lower in the PCA model

### Cluster Cryptocurrencies with K-means Using the Scaled PCA DataFrame

In [46]:
# Initialize the K-Means model using the best value for k
k=4
kmeans = KMeans(n_clusters=k, random_state=0)

In [47]:
# Fit the K-Means model using the PCA data
kmeans.fit(pca_df)

KMeans(n_clusters=4, random_state=0)

In [48]:
# Predict the clusters to group the cryptocurrencies using the scaled PCA DataFrame
clusters = kmeans.predict(pca_df)

# Print the resulting array of cluster values.
print("Cluster predictions:")
print(clusters)

Cluster predictions:
[2 2 0 0 2 2 2 2 2 0 0 0 0 2 0 2 0 0 2 0 0 2 0 0 0 0 0 0 2 0 0 0 1 2 0 0 3
 0 0 0 0]


In [49]:
# Create a copy of the scaled PCA DataFrame
pca_clustered_df = pca_df.copy()

# Add a new column to the copy of the PCA DataFrame with the predicted clusters
pca_clustered_df['cryptoclusters'] = clusters
pca_clustered_df['dataset'] = 'PCA Data'

# Display the copy of the scaled PCA DataFrame
pca_clustered_df.head()

PC1       PC2       PC3  cryptoclusters   dataset
coin_id                                                             
bitcoin      -0.600667  0.842760  0.461595               2  PCA Data
ethereum     -0.458261  0.458466  0.952877               2  PCA Data
tether       -0.433070 -0.168126 -0.641752               0  PCA Data
ripple       -0.471835 -0.222660 -0.479053               0  PCA Data
bitcoin-cash -1.157800  2.041209  1.859715               2  PCA Data

In [50]:
# Create a scatter plot using hvPlot by setting
# `x="PC1"` and `y="PC2"`.
# Color the graph points with the labels found using K-Means and
# add the crypto name in the `hover_cols` parameter to identify
# the cryptocurrency represented by each data point.
pca_scatter_plot = pca_clustered_df.hvplot.scatter(
    x="PC1",
    y="PC2",
    by="cryptoclusters",
    hover_cols=["coin_id"],
    title="PCA Cryptocurrency Clusters"
)
pca_scatter_plot

:NdOverlay   [cryptoclusters]
   :Scatter   [PC1]   (PC2,coin_id)

### Visualize and Compare the Results

In this section, you will visually analyze the cluster analysis results by contrasting the outcome with and without using the optimization techniques.

In [51]:
# Composite plot to contrast the Elbow curves
combined_elbow_df = pd.concat([elbow_df, elbow_df_pca])

composite_elbow_plot = combined_elbow_df.hvplot.line(
    x='k',
    y='inertia',
    by= 'Dataset',
    title='Elbow Curves: Scaled vs PCA Data',
    xlabel = 'Number of Clusters (k)',
    ylabel = 'Inertia'
)

composite_elbow_plot

:NdOverlay   [Dataset]
   :Curve   [k]   (inertia)

In [52]:
# Composite plot to contrast the clusters

composite_cluster_plot = scaled_scatter_plot + pca_scatter_plot
    
composite_cluster_plot

:Layout
   .NdOverlay.I  :NdOverlay   [CryptoCluster]
      :Scatter   [price_change_percentage_24h]   (price_change_percentage_7d,coin_id)
   .NdOverlay.II :NdOverlay   [cryptoclusters]
      :Scatter   [PC1]   (PC2,coin_id)

#### Answer the following question: 

  * **Question:** After visually analyzing the cluster analysis results, what is the impact of using fewer features to cluster the data using K-Means?

  * **Answer:** The clusters are more compact and concise leading to better categorization